packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -1634,3 +1671,148 @@ class _FifoCache<K, V> {
     return _cache[key] = loader();
   }
 }
+
+/// Defines the visual density of user interface components.
+///
+/// Density, in the context of a UI, is the vertical and horizontal
+/// "compactness" of the components in the UI. It is unitless, since it means
+/// different things to different UI components.
+///
+/// For example, for buttons, it affects the spacing around the child of the
+/// button. For lists, it affects the distance between baselines of entries in
+/// the list. For chips, it only affects the vertical size, not the horizontal
+/// size.
+class VisualDensity extends Diagnosticable {
+  /// A const constructor for [VisualDensity].
+  ///
+  /// All of the arguments must be non-null, and [horizontal] and [vertical]
+  /// must be in the interval between [minimumDensity] and [maximumDensity],
+  /// inclusive.
+  const VisualDensity({
+    this.horizontal = 0.0,
+    this.vertical = 0.0,
+    this.interval = 4.0,
+  })  : assert(horizontal != null),
```

Comment:
nit: 
```suggestion
  }) : assert(horizontal != null),
```
and all following assert statements